perf/perf.go

```diff
@@ -0,0 +1,520 @@
+			}
+		}
+	}()
+
+	// bpf_perf_event_output checks which CPU an event is enabled on,
+	// but doesn't allow using a wildcard like -1 to specify "all CPUs".
+	// Hence we have to create a ring for each CPU.
+	for i := 0; i < nCPU; i++ {
+		ring, err := newPerfEventRing(i, opts)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create perf ring for CPU %d", i)
+		}
+		rings = append(rings, ring)
+
+		if err := opts.Map.Put(uint32(i), uint32(ring.fd)); err != nil {
+			return nil, errors.Wrapf(err, "could't put event fd for CPU %d", i)
+		}
+
+		if err := addToEpoll(epollFd, ring.fd, int32(len(rings)-1), 0); err != nil {
+			return nil, err
+		}
+	}
+
+	closeFd, err := unix.Eventfd(0, unix.O_CLOEXEC|unix.O_NONBLOCK)
+	if err != nil {
+		return nil, err
+	}
+	fds = append(fds, closeFd)
+
+	if err := addToEpoll(epollFd, closeFd, 0, wokenByClose); err != nil {
+		return nil, err
+	}
+
+	array, err := opts.Map.Clone()
+	if err != nil {
+		return nil, err
+	}
+
+	pr = &Reader{
+		array:   array,
+		rings:   rings,
+		epollFd: epollFd,
+		// Allocate extra event for closeFd
+		epollEvents: make([]unix.EpollEvent, len(rings)+1),
+		epollRings:  make([]*perfEventRing, 0, len(rings)),
+		closeFd:     closeFd,
+	}
+	runtime.SetFinalizer(pr, (*Reader).Close)
+	return pr, nil
+}
+
+// Close frees resources used by the reader.
+//
+// It interrupts calls to Read.
+//
+// Calls to perf_event_output from eBPF programs will return
+// ENOENT after calling this method.
+func (pr *Reader) Close() error {
+	var err error
+	pr.closeOnce.Do(func() {
+		runtime.SetFinalizer(pr, nil)
+
+		// Interrupt Read() via the event fd.
+		var value [8]byte
+		internal.NativeEndian.PutUint64(value[:], 1)
+		_, err = unix.Write(pr.closeFd, value[:])
+		if err != nil {
+			err = errors.Wrap(err, "can't write event fd")
+			return
+		}
+
+		// Acquire the lock. This ensures that Read
+		// isn't running.
+		pr.mu.Lock()
+		defer pr.mu.Unlock()
+
+		unix.Close(pr.epollFd)
+		unix.Close(pr.closeFd)
+		pr.epollFd, pr.closeFd = -1, -1
+
+		// Close rings
+		for _, ring := range pr.rings {
+			ring.Close()
+		}
+		pr.rings = nil
+
+		pr.array.Close()
+	})
+
+	return errors.Wrap(err, "close PerfReader")
+}
+
+// Read the next record from the perf ring buffer.
+//
+// The function blocks until there are at least Watermark bytes in one
+// of the per CPU buffers.
+//
+// Records from buffers below the Watermark are not returned.
+//
+// Calling Close interrupts the function.
+func (pr *Reader) Read() (*Record, error) {
+	pr.mu.Lock()
+	defer pr.mu.Unlock()
+
+	if pr.epollFd == -1 {
+		return nil, errors.New("closed")
+	}
+
+	for {
+		if len(pr.epollRings) == 0 {
+			nEvents, err := unix.EpollWait(pr.epollFd, pr.epollEvents, -1)
+
+			// Handle EINTR
+			if temp, ok := err.(temporaryError); ok && temp.Temporary() {
+				continue
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			for _, event := range pr.epollEvents[:nEvents] {
+				if event.Pad == wokenByClose {
```

Comment:
Do we need to use `Pad` and `Fd` this way? I find it confusing to follow that `event.Fd` has the CPU number. What if we put actual fds in `Fd`, and have a `map[fd]ring` and check `event.Fd == closeFd`?

Reply:
True, but in a way that is also obfuscating what is going on. There is no requirement for event to contain the fd, its just that `unix` named it so, which is why I added that hard to understand comment about the union further up.

There is also a bit of premature optimization here, because lookup in small maps are "slow":  https://github.com/golang/go/issues/19495 By using the CPU here we can zero index into a slice instead of using a map. Not sure if it really makes a difference.

I'll have a think whether I can keep using event like this but make it less confusing.

Reply:
I changed it around a little bit: `event.Fd` now really does contain the fd and we compare against `closeFd`. CPU has moved to event.Pad, and there is a small accessor function `cpuForEvent` which makes the code easier to read.